### PR TITLE
planner, ddl: convert large varchar for create table in non strict mode

### LINKED ISSUE #30328: Create table with column length is too big : Mysql is converting VARBINARY to BLOB and VARCHAR to TEXT , Tidb is resulting error.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
SET sql_mode = ‘NO_ENGINE_SUBSTITUTION’;
drop table if exists t1;
CREATE TABLE t1( a varbinary(70000), b varchar(70000000));
desc t1;
```
### 2. What did you expect to see? (Required)
```sql
mysql> CREATE TABLE t1( a varbinary(70000), b varchar(70000000));
Query OK, 0 rows affected, 2 warnings (0.01 sec)

mysql> desc t1;
+-------+------------+------+-----+---------+-------+
| Field | Type       | Null | Key | Default | Extra |
+-------+------------+------+-----+---------+-------+
| a     | mediumblob | YES  |     | NULL    |       |
| b     | longtext   | YES  |     | NULL    |       |
+-------+------------+------+-----+---------+-------+
2 rows in set (0.00 sec)
```
### 3. What did you see instead (Required)
```sql
mysql> CREATE TABLE t1( a varbinary(70000), b varchar(70000000));
ERROR 1074 (42000): Column length too big for column 'a' (max = 65535); use BLOB or TEXT instead
```
### 4. What is your TiDB version? (Required)
```sql
| Release Version: v5.4.0-alpha-264-g6efa36df6
Edition: Community
Git Commit Hash: 6efa36df6cff325106f67ecfe3d79816ba37ca6a
Git Branch: master
UTC Build Time: 2021-11-29 16:57:51
GoVersion: go1.17.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
```

### PR DESCRIPTION
**Pull request by unconsolable**
Signed-off-by: unconsolable <chenzhipeng2012@gmail.com>

### What problem does this PR solve?

Issue Number: close #30328 

Problem Summary:

### What is changed and how it works?
- In non strict mode, auto convert varchar with large length to TEXT or BLOB, and append warnings.
- Fix a nil pointer dereference in test case due to changes.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by unconsolable** [comment]
/cc @djshow832 @tangenta 
PTAL, thanks

File: planner/core/preprocess_test.go [hunk 1]
```diff
@@ -364,3 +364,18 @@ func (s *testValidatorSuite) TestDropGlobalTempTable(c *C) {
...
+func (s *testValidatorSuite) TestLargeVarcharAutoConv(c *C) {
+	_, err := s.se.Execute(context.Background(), "use test")
+	c.Assert(err, IsNil)
+	s.runSQL(c, "CREATE TABLE t1(a varbinary(70000), b varchar(70000000))", false,
+		errors.New("[types:1074]Column length too big for column 'a' (max = 65535); use BLOB or TEXT instead"))
+
+	_, err = s.se.Execute(context.Background(), "SET sql_mode = 'NO_ENGINE_SUBSTITUTION'")
+	c.Assert(err, IsNil)
+	s.runSQL(c, "CREATE TABLE t1(a varbinary(70000), b varchar(70000000));", false, nil)
+	s.runSQL(c, "CREATE TABLE t1(a varbinary(70000), b varchar(70000000) charset utf8mb4);", false, nil)
```
Comment (sylzd) [inline planner/core/preprocess_test.go:378]: can you add warnings test here?

Comment (unconsolable) [reply planner/core/preprocess_test.go:378]: Resolved.

**Review by sylzd — APPROVED** [review]

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -587,6 +587,22 @@ func buildColumnAndConstraint(
 	if err := setCharsetCollationFlenDecimal(colDef.Tp, chs, coll); err != nil {
 		return nil, nil, errors.Trace(err)
 	}
+	// Some column checks are left here, so varchar auto convert is performed. See also issue 30328
+	sessVars := ctx.GetSessionVars()
+	if sessVars != nil && !sessVars.SQLMode.HasStrictMode() && colDef.Tp.Tp == mysql.TypeVarchar {
```
Comment (unconsolable) [inline ddl/ddl_api.go:592]: I wonder whether here needs a test too. If it is, where shall I add the test? (`db_xxx.go` is a bit confusing)

**Comment by unconsolable** [comment]
/cc @bb7133 
PTAL, thanks

**Comment by unconsolable** [comment]
I haven't take a look of this PR for a while. I found that it has some other problems with current code, but I don't have time to fix it now....

**Comment by djshow832** [comment]
@unconsolable  Sorry that I missed the message. Which kind of problem did you meet? Code conflicts or bugs?

**Comment by unconsolable** [comment]
> 

It fails on some other tests, I was a bit busy at that time, and I can't take time on fixing it, so I closed the issue and unassigned myself.

**Comment by unconsolable** [comment]
> @unconsolable Sorry that I missed the message. Which kind of problem did you meet? Code conflicts or bugs?

It fails on some other tests, I was a bit busy at that time, and I can't take time on fixing it, so I closed the issue and unassigned myself.

**Pull request closed without merging**